[FEATURE][needs-docs] Add gui to draw Circle, Ellipse, Rectangle and Regular Polygon

lbartoletti opened the pull request:
## Description
Since classes are available in the core, I have taken the time to write a GUI to draw these shapes. 

### Discussion
- As said on the maillist, there is a class by tool into a new toolbar called mRegularShapeDigitizeToolBar. Maybe it can be renamed to mShapeDigitizeToolBar since a rectangle is not regular (except square...).
- Rectangle:
    - I use QgsRectangle class. I think, it's not a good solution because of some limitations of this class, mostly used to be a bounding box. I don't write a class like QgsRegularPolygon but it's maybe a better solution where we can really use it to draw oriented rectangle (3 points; a point with height, width, azimuth; add inclinations; etc.) or maybe rewrite this class heavily.
    - I have written a tool for square, but I think it's useless since it's a regular polygon and can be constructed as it.

- Regular polygon:
    - I created an input user to set the number of vertices. Is like for circular string, but personally I don't like it. I prefer, if it's possible, to add this widget into the toolbar (like with CADDigitize plugin).
    - The number of sides is by default the minimum, but do you think the last value used must be stored and reused?

### New tools
- Circle from 2 points: Points define the diameter and the orientation of the circle.
- Circle from 3 points
- Circle from center and a point: with cadDockWidget can be "center and a radius" using the distance tool
- Circle from 3 tangents: you have to activate snapping segment. Obviously, you must click on a segment to add a tangent. If two tangents are parallels a message appear and clear input.
- Circle from 2 tangents: As circle from 3 tangents, except that you have to select to tangents, enter a radius and select the center. This tool is the same as QCad and not as AutoCAD.
- Ellipse from center and 2 points: The first point defines the orientation and the length of the first axis. The second point on the ellipse defines the length of the second axis.
- Ellipse from center and a point: Like an ellipse drawn into a bounding box with center and a corner.
- Ellipse from extent.
- Ellipse from foci: 2 points for foci and a point on the ellipse.
- Rectangle from center and a point.
- Rectangle from extent.
- Rectangle from 3 points.
- Square from center. Useless, can be regular polygon from center.
- Regular polygon from center (and a corner): The corner defines the orientation of the regular polygon.
- Regular polygon from 2 corners.

Also, circles are sending to circular string so it can be stored into a curve or segmented for other geometry.
All tools are in the toolbar and menu (and submenu).
Since doc is not in this repository, I haven't written it, but it can easily done using this information, docs from CADDigitize and generally speaking from all CAD tools.

### Modifications of other class:
- QgsCircle: Fix error for QgsCircle::from2Points. I used diameter instead of radius...
- QgsEllipse and QgsRegularPolygon: close linestring
- QgsMapToolCapture: add a convenient method to clear curve, especially used for these tools.
- QgsTriangle: I have added new tests to QgsTriangle since I have added preconditions for empty or invalid triangles.

Regards.


## Checklist

> Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

- [X] Commit messages are descriptive and explain the rationale for changes
- [ ] Commits which fix bugs include `fixes #11111` in the commit message next to the description
- [X] Commits which add new features are tagged with `[FEATURE]` in the commit message
- [X] Commits which change the UI or existing user workflows are tagged with `[needs-docs]` in the commit message and containt sufficient information in the commit message to be documented
- [X] I have read the [QGIS Coding Standards](https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
- [X] This PR passes all existing unit tests (test results will be reported by travis-ci after opening this PR)
- [x] New unit tests have been added for core changes
- [X] I have run [the `scripts/prepare-commit.sh` script](https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTE.md#contributing-to-qgis) before each commit

lbartoletti committed:
- Init GUI for RegularShapes (add a Toolbar:
RegularShapeDigitizeToolBar)
- Add Circle Tools (2 Points on circle, 3 Points on circle, center and a
point)
- Fix an error on QgsCircle

lbartoletti committed:
- Add ellipse tools
- Change QgsEllipse().toLineString to have closed linestring

lbartoletti committed:
Add rectangle and regular polygon.

Rectangle crash

lbartoletti committed:
Fix tests

lbartoletti committed:
Fix crash for rectangle (centerPoint - extent)

lbartoletti committed:
Remove rectangle3points (crash)

lbartoletti committed:
rm qgsmaptooladdregularshape.{cpp,h}

lbartoletti committed:
clean

lbartoletti committed:
Merge branch 'master' into regularShape_gui

lbartoletti committed:
Add CircleExtent svg

lbartoletti committed:
Shapes drawing with 3 points are no longer frozen when a left click is
made for the 3rd point.

lbartoletti committed:
Add circle from 3 tangents

lbartoletti committed:
- Fix crash for QgsMapToolCircle3Tangents when segment is not activated
or segments are parallels.
- Add preconditions for methods of QgsTriangle when is empty or invalid.
Tests and docs are updated.

lbartoletti committed:
Merge branch 'master' into regularShape_gui

lbartoletti commented:
Fails since  #5209 that was merged.

lbartoletti committed:
Merge branch 'master' of https://github.com/QGIS/QGIS into regularShape_gui

lbartoletti committed:
Fix build with improvement of #5209

lbartoletti committed:
Fix error when the tools was changed. setParentTool is not needed.

lbartoletti committed:
Fix remaining rubberBand

lbartoletti committed:
Add Rectangle from 3 points

lbartoletti committed:
Merge branch 'master' into regularShape_gui

lbartoletti committed:
Use unique_ptr

lbartoletti committed:
Merge branch 'regularShape_gui' of https://github.com/lbartoletti/QGIS into regularShape_gui

lbartoletti commented:
A tiny demo (click for the video):
[![Demo](https://raw.githubusercontent.com/lbartoletti/lbartoletti.github.io/master/example.png)](https://raw.githubusercontent.com/lbartoletti/lbartoletti.github.io/master/example.webm)

nyalldawson commented:
@lbartoletti 
Please don't be discouraged by the lack of responses here so far - these are very valuable improvements and I'm excited to see it merged. 

I just haven't had a chance to review yet, but will ASAP.

lbartoletti commented:
@nyalldawson No problem! I haven't totally finished the tools and I have found a bug today.

lbartoletti committed:
Fix tools with extent when map is rotated (add a messagebar)

lbartoletti committed:
resolve conflicts

lbartoletti committed:
Merge branch 'master' into regularShape_gui

lbartoletti committed:
Fix minimum for QgsMapToolRegularPolygonCenterPoint

lbartoletti committed:
Suppress old constructors

lbartoletti committed:
Remove no more usable tools Circle by 3 tangents

lbartoletti committed:
Fix tools which doesn't works when map is rotated

lbartoletti committed:
Merge branch 'master' into regularShape_gui

lbartoletti committed:
Modernize

lbartoletti commented:
It works (c)
:black_square_button: :large_blue_circle: :large_orange_diamond: :small_red_triangle_down:

lbartoletti committed:
Reborn of QgsMapToolCircle3Tangents. Thanks Martin!

lbartoletti committed:
- Add QgsMapToolRegularPolygonCenterCorner and change
QgsMapToolRegularPolygonCenterPoint
- Add QgsMapToolCircle2TangentsPoint

lbartoletti committed:
Merge branch 'master' into regularShape_gui

pblottiere commented on src/app/qgsmaptooladdellipse.cpp:
```diff
@@ -0,0 +1,96 @@
+QgsMapToolAddEllipse::QgsMapToolAddEllipse( QgsMapToolCapture *parentTool, QgsMapCanvas *canvas, CaptureMode mode )
+  : QgsMapToolCapture( canvas, QgisApp::instance()->cadDockWidget(), mode )
+  , mParentTool( parentTool )
+  , mTempRubberBand( nullptr )
```
Already initialized in header

lbartoletti replied:
OK done for all classes

pblottiere commented on src/app/qgsmaptooladdellipse.cpp:
```diff
@@ -0,0 +1,96 @@
+  if ( mTempRubberBand )
+  {
+    delete mTempRubberBand;
+    mTempRubberBand = nullptr;
```
Do you think that `std::unique_ptr` could be used for `mTempRubberBand`?

lbartoletti replied:
It wasn't not used in `QgsMapToolCircularString`.
I think that if we want use `std::unique_ptr` we have to make other changes in `QgsMapToolEdit` because we use `createGeometryRubberBand` which allow to have an alternativeBand like.
Maybe we have to change all other classes using rubberband. Another PR?

pblottiere commented on src/app/qgsmaptooladdregularpolygon.cpp:
```diff
@@ -0,0 +1,121 @@
+  {
+    QgisApp::instance()->statusBarIface()->removeWidget( mNumberSidesSpinBox );
+    delete mNumberSidesSpinBox;
+    mNumberSidesSpinBox = nullptr;
```
Maybe `std::unique_ptr` to avoid the manual deletion?

lbartoletti replied:
Done.
Is std::make_unique (c++14) will be usable?
Wasn't for me.

nyalldawson replied:
No, not available in master unfortunately

lbartoletti committed:
Fix qgisapp.ui

pblottiere commented on src/app/qgsmaptooladdregularpolygon.h:
```diff
@@ -0,0 +1,49 @@
+#ifndef QGSMAPTOOLADDREGULARPOLYGON_H
```
The header (copyright, email, ...) seems to be lacking for this file.

lbartoletti replied:
done

pblottiere commented on src/app/qgsmaptoolcircle2points.cpp:
```diff
@@ -0,0 +1,64 @@
+
+    if ( !mPoints.isEmpty() )
+    {
+      if ( !mTempRubberBand )
```
Another minor comment : these two successives `if` may be concatenated to avoid an unneeded level of indentation : 

```
if ( !mPoints.isEmpty() && !mTempRubberBand )
```
```

lbartoletti replied:
done

pblottiere commented on src/app/qgsmaptooladdrectangle.cpp:
```diff
@@ -0,0 +1,163 @@
+  QgsPoint x1, x2, x3;
+  if ( isOriented )
+  {
+    double perpendicular = 90.0 * mSide;
```
`const` value

lbartoletti replied:
done

pblottiere commented on src/app/qgsmaptoolcircle2tangentspoint.cpp:
```diff
@@ -0,0 +1,274 @@
+    double currentDist = mapPoint.distanceSquared( center );
+    for ( int i = 1; i < mCenters.size(); ++i )
+    {
+      double testDist = mapPoint.distanceSquared( mCenters.at( i ).x(), mCenters.at( i ).y() );
```
`const` value

lbartoletti replied:
done

pblottiere commented on src/app/qgsmaptoolregularpolygoncentercorner.h:
```diff
@@ -0,0 +1,34 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
```
Header should be on top of the file... But maybe it doesn't matter?

lbartoletti replied:
done for all headers

pblottiere commented on src/app/qgsmaptoolregularpolygoncenterpoint.h:
```diff
@@ -0,0 +1,34 @@
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ ***************************************************************************/
```
Header on top?

lbartoletti replied:
done

pblottiere commented on src/core/geometry/qgstriangle.h:
```diff
@@ -246,85 +311,134 @@ class CORE_EXPORT QgsTriangle : public QgsPolygonV2
-    * \endcode
-    */
+     * Circumscribed circle of the triangle.
+     * @return The circumbscribed of the triangle with a QgsCircle.
```
I think it should be `\returns` instead of `@return`?

lbartoletti replied:
done

pblottiere commented on src/core/geometry/qgstriangle.h:
```diff
@@ -246,85 +311,134 @@ class CORE_EXPORT QgsTriangle : public QgsPolygonV2
-    * \endcode
-    */
+     * Inscribed circle of the triangle.
+     * @return The inscribed of the triangle with a QgsCircle.
```
`\returns`?

lbartoletti replied:
done

pblottiere commented:
Nice work @lbartoletti! Just some minor comments.

pblottiere commented:
@lbartoletti 

I didn't succeed in using the `Add circle from 3 Tangents` map tool.

Moreover I had a segmentation fault when I was using `QgsMapToolCircle2TangentsPoint` : https://gist.github.com/pblottiere/fc91e77ee86253e775e0d49c8f7d9552

Apart from that, other map tools seem to work very well!

lbartoletti commented:
Thanks @pblottiere for your reviewing!

For circle from 3 Tangents you have to enable snapping and snap on segments. Also, it's better to use these two tools with cad dock enabled.

QgsMapToolCircle2TangentsPoint needs to be cleaned. I will check it.

Thanks again.

lbartoletti committed:
Remove initialization when it's already in header

lbartoletti committed:
Add missing copyright for QgsMapToolAddRegularPolygon

lbartoletti committed:
Move header and normalize space

lbartoletti committed:
@return -> \returns

lbartoletti committed:
Add const

lbartoletti commented on src/app/qgsmaptoolcircle2tangentspoint.cpp:
```diff
@@ -0,0 +1,274 @@
+    }
+
+    QgsPoint center = QgsPoint( mCenters.at( 0 ) );
+    double currentDist = mapPoint.distanceSquared( center );
```
done

lbartoletti committed:
Concat successive if

lbartoletti committed:
Concat successive if

lbartoletti committed:
Merge branch 'regularShape_gui' of https://github.com/lbartoletti/QGIS into regularShape_gui

lbartoletti committed:
modernize (delete default) destructor

lbartoletti committed:
Use unique_ptr for mNumberSidesSpinBox

lbartoletti committed:
Visual indent part 1

lbartoletti committed:
Visual indent part 2

lbartoletti committed:
Visual indent part 3

lbartoletti committed:
Uniformize message when segments are parallels

lbartoletti committed:
Change rubberband of possible centers to point.

lbartoletti committed:
remove SquareCenter use RegularPolygon instead

nyalldawson commented on src/app/qgsmaptoolcircle2tangentspoint.cpp:
```diff
@@ -0,0 +1,271 @@
+void QgsMapToolCircle2TangentsPoint::createRadiusSpinBox()
+{
+  deleteRadiusSpinBox();
+  mRadiusSpinBox = new QSpinBox();
```
QgsSpinBox

nyalldawson commented on src/app/qgsmaptoolcircle2tangentspoint.cpp:
```diff
@@ -0,0 +1,271 @@
+{
+  if ( mRadiusSpinBox )
+  {
+    QgisApp::instance()->statusBarIface()->removeWidget( mRadiusSpinBox );
```
again, is this correct?

nyalldawson commented on src/app/qgsmaptoolcircle2tangentspoint.cpp:
```diff
@@ -0,0 +1,271 @@
+  mRadius = radius;
+  getPossibleCenter( );
+
+  for ( int i = 0; i < mRubberBands.size() ; ++i )
```
change this to just qDeleteAll( mRubberBands )

nyalldawson commented on src/app/qgsmaptoolcircle2tangentspoint.h:
```diff
@@ -0,0 +1,59 @@
+    ~QgsMapToolCircle2TangentsPoint();
+
+    void cadCanvasReleaseEvent( QgsMapMouseEvent *e ) override;
+    void cadCanvasMoveEvent( QgsMapMouseEvent *e );
```
override

nyalldawson commented on src/app/qgsmaptoolcircle2tangentspoint.h:
```diff
@@ -0,0 +1,59 @@
+    void radiusSpinBoxChanged( int radius );
+
+  private:
+    //! Return the point where segments are intersected. Method from QgsGeometryUtils doesn't work for special cases used by this tool.
```
do you think the QgsGeometryUtils method should be fixed?

lbartoletti replied:
Not at the moment. I have launched a QEP to improve editing tools and some of these will need this kind of method. The QgsGeometryUtils' method is used for the topology. I have found (at the moment) only one case where there is a problem (for this intersection): when lines are continuous.A rewrite can will be done by merging this needs and the other editing tools.

nyalldawson commented on src/app/qgsmaptoolellipsecenter2points.cpp:
```diff
@@ -0,0 +1,83 @@
+/***************************************************************************
+    qgmaptoolellilpsecenter2points.cpp  -  map tool for adding ellipse
```
typo

nyalldawson commented on src/app/qgsmaptoolellipseextent.cpp:
```diff
@@ -0,0 +1,80 @@
+        }
+        else
+        {
+          emit messageEmitted( tr( "Cannot use this tool when the map canvas is rotated" ), QgsMessageBar::WARNING );
```
I'm curious about this - couldn't we rotate the points /unrotate the result to avoid this limitation?

lbartoletti replied:
At the moment, I haven't been able to make it.

nyalldawson commented on src/app/qgsmaptoolrectanglecenter.cpp:
```diff
@@ -0,0 +1,87 @@
+        }
+        else
+        {
+          emit messageEmitted( tr( "Cannot use this tool when the map canvas is rotated" ), QgsMessageBar::WARNING );
```
again, would rotating/unrotating work?

lbartoletti replied:
After having revised my trigonometry and reinvented the formulas, I realized that it was obvious.... I was blind...

It's fixed

nyalldawson commented on src/app/qgsmaptoolregularpolygoncentercorner.cpp:
```diff
@@ -0,0 +1,79 @@
+
+QgsMapToolRegularPolygonCenterCorner::~QgsMapToolRegularPolygonCenterCorner()
+{
+  if ( mNumberSidesSpinBox )
```
no need for an if here

nyalldawson commented on src/app/qgsmaptoolregularpolygoncenterpoint.cpp:
```diff
@@ -0,0 +1,79 @@
+
+QgsMapToolRegularPolygonCenterPoint::~QgsMapToolRegularPolygonCenterPoint()
+{
+  if ( mNumberSidesSpinBox )
```
again, redundant if (maybe elsewhere in destructors too)

nyalldawson commented:
Phenomenal work @lbartoletti ! This is going to be such a nice feature to have in master.

Aside from the code notes above, I've hit a couple of issues while testing:

1. The old "add circular string" tool seems broken when used on a polygon layer. I couldn't reproduce in master, so I assume it's something introduced in this PR

2. when digitizing the rubber bands are showing all the extra nodes coming from the segmentization used in the rubber band. Is this just a result of missing methods to convert QgsCircles/QgsEllipses to curve polygons? it'd be nice to avoid these as they add a lot of extra "visual noise"

3. something's not quite right in the "add circle" tools. Try adding a polygon memory layer and creating a circle in using this tool, then zooming in. The result isn't a circle :)

4. I think the ellipse tools need more segments by default - the results are often quite jagged

5. after finishing digitizing a shape, the tool (in the toolbar) gets deselected. but it's still active in the canvas.

lbartoletti committed:
Requests for QgsMapToolAddCircle

lbartoletti commented on src/app/qgsmaptooladdcircle.h:
```diff
@@ -0,0 +1,59 @@
+    //! Circle points (in map coordinates)
+    QgsPointSequence mPoints;
+    //! The rubberband to show the circular string currently working on
+    QgsGeometryRubberBand *mTempRubberBand = nullptr;
```
As said as @pblottiere ([link](https://github.com/qgis/QGIS/pull/4947#discussion_r142491771)), I think there is other changes to did it.
I've made a test without success at the moment.

lbartoletti commented on src/app/qgsmaptooladdellipse.cpp:
```diff
@@ -0,0 +1,98 @@
+  }
+
+  mParentTool->clearCurve();
+  mParentTool->addCurve( mEllipse.toLineString() );
```
I haven't found a way or math article to translate an ellipse to a circularstring. Is what is not in QgsEllipse class.

nyalldawson replied:
Hmm... true. Seems like a huge shortcoming in the curved geometry model!

lbartoletti replied:
Yes. I have checked today what is done with AutoCAD. It's also a polyline approximation.
I will try to find the best approximation later, that's what he seems to be made for Cairo.

lbartoletti committed:
- mParentTool -> intialize to nullptr;
- =delete to explicit class;
- move clean() to private;
- remove unnecessary initializer;

lbartoletti commented on src/app/qgsmaptooladdregularpolygon.cpp:
```diff
@@ -0,0 +1,123 @@
+{
+  if ( mNumberSidesSpinBox )
+  {
+    QgisApp::instance()->statusBarIface()->removeWidget( mNumberSidesSpinBox.get() );
```
No. Deleted.

lbartoletti committed:
Requests for QgsMapToolAddRegularPolygon

lbartoletti committed:
Requests for QgsMapToolCircle2TangentsPoint

lbartoletti committed:
Typo for qgsmaptoolellipsecenter2points.{cpp,h}

lbartoletti committed:
- Request for qgsmaptoolregularpolygoncentercorner.cpp and
qgsmaptoolregularpolygoncenterpoint.cpp
- Remove useless mNumberSidesSpinBox->setMinimum( 3 );

lbartoletti commented:
Thank you @nyalldawson !
To answer your points:
1. I have reproduced it. At the moment, I have no idea of why it happens.
2. QgsCircle have a method to curve polygon, but not QgsEllipse. As I wrote earlier, I don't know a method to convert QgsEllipse to a circular string. For QgsCircle it's fixed.
3. I have reproduced it but with a Geographic Coordinate System (WGS84). For Projected Coordinate System it's ok.
4. My dream is to find a method to convert an ellipse to a circular sting. Otherwise, there are other possibilities: retrieve the existing option in QGis, add a widget (what there was in CAD Digitize), automatically define a segmentation according to the size of the largest axis (I don't know any method for that, I think it's to be found empirically.) It has to be simple for the user, that's why I didn't apply method 2 in QGis. I'm gonna run some tests with Method 1 and 3. If I find a solution, I'll propose another PR. In the meantime, we can already validate the increase of the number of segments. 
5. You're right. I will check it.

@pblottiere I can't reproduce you crash with `QgsMapToolCircle2TangentsPoint`

I have found another bug:
The rubber band remains when you change tool during using it.

lbartoletti committed:
Use toCircularString instead of toPolygon for circle.

lbartoletti committed:
Fix indent

lbartoletti committed:
Improve visibility of tangents tool on cadCanvasMoveEvent (compatible
when CADDockWidget is enabled)

nyalldawson commented:
> 1. I have reproduced it. At the moment, I have no idea of why it happens.

Ok - this will need to be fixed before we can merge this PR.

> 2 QgsCircle have a method to curve polygon, but not QgsEllipse. As I wrote earlier, I don't know a method to convert QgsEllipse to a circular string. For QgsCircle it's fixed.

Thanks! It indeed seems that there's no way to represent ellipses using curved geometries... it's a shame, but nothing we can address.

> 3. I have reproduced it but with a Geographic Coordinate System (WGS84). For Projected Coordinate System it's ok.

So a problem with rendering reprojected curved geometries?
 
> 4.My dream is to find a method to convert an ellipse to a circular sting. Otherwise, there are other possibilities: retrieve the existing option in QGis, add a widget (what there was in CAD Digitize), automatically define a segmentation according to the size of the largest axis (I don't know any method for that, I think it's to be found empirically.) It has to be simple for the user, that's why I didn't apply method 2 in QGis. I'm gonna run some tests with Method 1 and 3. If I find a solution, I'll propose another PR. In the meantime, we can already validate the increase of the number of segments.

Sounds like a good approach.

> 5. You're right. I will check it.

Thanks!

lbartoletti committed:
- Increases the number of segments for QgsMapToolAddEllipse

lbartoletti commented:
>     2 QgsCircle have a method to curve polygon, but not QgsEllipse. As I wrote earlier, I don't know a method to convert QgsEllipse to a circular string. For QgsCircle it's fixed.
> 
> Thanks! It indeed seems that there's no way to represent ellipses using curved geometries... it's a shame, but nothing we can address.

It's fixed for QgsEllipse (00e12bc).

>     3 I have reproduced it but with a Geographic Coordinate System (WGS84). For Projected Coordinate System it's ok.
> 
> So a problem with rendering reprojected curved geometries?

Maybe, see the screenshot. I have draw a circle. The geometry (purple) is strange, but the identified geometry (red) is ok. 
<a href="https://ibb.co/b4a4Tw"><img src="https://thumb.ibb.co/b4a4Tw/Screenshot_20171009_231226.png" alt="Screenshot_20171009_231226" border="0"></a>

lbartoletti committed:
remove setParentTool method in QgsMapToolAddCircularString

lbartoletti commented:
@nyalldawson FYI, I have open two bug reports for [3](https://issues.qgis.org/issues/17269) and [5.](https://issues.qgis.org/issues/17270)

lbartoletti committed:
Use lambdas in the connect for tools (and circular tools)

lbartoletti committed:
Avoid duplicate code for check if the layer is editable

lbartoletti commented on src/app/qgisapp.cpp:
```diff
@@ -7678,6 +7811,81 @@ void QgisApp::circularStringRadius()
   mMapCanvas->setMapTool( mMapTools.mCircularStringRadius );
 }
 
+void QgisApp::circle2Points()
```
I have made a function and replace by lambdas in the connect.
Other tools must be convert to this, if you accept the solutions. For now, I have fixed only my commit and circular string tools.

nyalldawson replied:
Looks perfect - much nicer!

lbartoletti committed:
Change QgsMapTool *mAddFeature to QgsMapToolAddFeature *mAddFeature and
remove dynamic_cast

lbartoletti committed:
Fix remaining rubber band

lbartoletti committed:
Enable tools using extent when map is rotated

lbartoletti committed:
manual resolve conflicts for qgisapp.ui

nyalldawson commented:
@lbartoletti what's the latest status here? It seems to me that you've addressed everything I pointed out now?

lbartoletti commented:
@nyalldawson Yes. Just leave me the weekend to do one last check.

lbartoletti committed:
- Add Signals start/stopCapturing
- fix QSpinBox -> QgsSpinBox

lbartoletti commented:
Go? :)

nyalldawson commented:
Yes please!! 

Thanks for your tremendous work on this... it's going to be very much appreciated amongst our users!

pblottiere commented:
Thank you @lbartoletti! Amazing work!

saberraz commented:
Thanks a lot @lbartoletti 
A quick comment: the toolbar for the new tool appears as "toolBar" in the list. I think, it should be a part of the Advanced Digitizing toolbar (my preferred option) or a new toolbar.

haubourg commented:
\°/ 
bravo Loïc!

3nids commented on src/ui/qgisapp.ui:
```diff
@@ -609,6 +638,17 @@
   </widget>
+  <widget class="QToolBar" name="mRegularShapeDigitizeToolBar">
+   <property name="windowTitle">
+    <string>toolBar</string>
```
@lbartoletti the name of the toolbar is not specified.
Also, I'm not sure it makes sense to add a new tool bar, I would rather integrate it in the digitizing toolbar (since you can customize it), but not 100% convinced.

lbartoletti commented:
Thank you, and especially thanks to @pblottiere and @nyalldawson for their reviews and advice!

Goodbye CADDigitize, long life to QGis!

@3nids and @saberraz, It may indeed make more sense to group tools together with CircularString. Especially that it would be necessary in the long term to add the other possibilities of creating arcs (see CADDigitize and QAD) and elliptical arcs. I will open a PR with your suggestion.
